synapse/lib/stormlib/auth.py

```diff
@@ -88,12 +127,234 @@
+                auth.role.delrule ninjas --index  5
+        ''',
+        'cmdargs': (
+            ('name', {'type': 'str', 'help': 'The name of the role.'}),
+            ('rule', {'type': 'str', 'help': 'The rule string.'}),
+            ('--gate', {'type': 'str', 'help': 'The auth gate id to grant permission on.', 'default': None}),
+            ('--index', {'type': 'bool', 'action': 'store_true', 'default': False,
+                'help': 'Specify the rule as a 0 based index into the list of rules.'}),
+        ),
+        'storm': '''
+            $role = $lib.auth.roles.byname($cmdopts.name)
+            if $role {
+                if $cmdopts.index {
+                    $rule = $role.popRule($cmdopts.rule, gateiden=$cmdopts.gate)
+                } else {
+                    $rule = $lib.auth.ruleFromText($cmdopts.rule)
+                    $role.delRule($rule, gateiden=$cmdopts.gate)
+                }
+                $ruletext = $lib.auth.textFromRule($rule)
+                $lib.print(`Removed rule {$ruletext} from role {$cmdopts.name}.`)
+            } else {
+                $lib.warn(`Role ({$cmdopts.name}) not found!`)
+            }
+        ''',
+    },
+    {
+        'name': 'auth.user.grant',
+        'descr': '''
+            Grant a role to a user.
+
+            Examples:
+                // Grant the role "ninjas" to the user "visi"
+                auth.user.grant visi ninjas
+
+        ''',
+        'cmdargs': (
+            ('username', {'type': 'str', 'help': 'The name of the user.'}),
+            ('rolename', {'type': 'str', 'help': 'The name of the role.'}),
+        ),
+        'storm': '''
+            $user = $lib.auth.users.byname($cmdopts.username)
+            if (not $user) { $lib.exit(`No user named: {$cmdopts.username}`) }
+
+            $role = $lib.auth.roles.byname($cmdopts.rolename)
+            if (not $role) { $lib.exit(`No role named: {$cmdopts.rolename}`) }
+
+            $lib.print(`Granting role {$role.name} to user {$user.name}.`)
+            $user.grant($role.iden)
+        ''',
+    },
+    {
+        'name': 'auth.user.revoke',
+        'descr': '''
+            Revoke a role from a user.
+
+            Examples:
+                // Revoke the role "ninjas" from the user "visi"
+                auth.user.grant visi ninjas
+
+        ''',
+        'cmdargs': (
+            ('username', {'type': 'str', 'help': 'The name of the user.'}),
+            ('rolename', {'type': 'str', 'help': 'The name of the role.'}),
+        ),
+        'storm': '''
+            $user = $lib.auth.users.byname($cmdopts.username)
+            if (not $user) { $lib.exit(`No user named: {$cmdopts.username}`) }
+
+            $role = $lib.auth.roles.byname($cmdopts.rolename)
+            if (not $role) { $lib.exit(`No role named: {$cmdopts.rolename}`) }
+
+            if (not $user.roles.has($role.iden)) {
+                $lib.exit(`User {$cmdopts.username} does not have role {$cmdopts.rolename}`)
```

Comment:
the `del.rule` commands do not call lib.exit when the rule does not exist on a user. it feels a bit weird for the revoke command to throw a StormExit when the user doesn't have the role?

Reply:
per discussion we'll leave this as is